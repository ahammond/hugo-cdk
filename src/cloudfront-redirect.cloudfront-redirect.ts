import * as awsLambda from 'aws-lambda';

function generateRedirectResponse(value: string): awsLambda.CloudFrontRequestResult {
  return {
    status: '301',
    statusDescription: 'Found',
    headers: {
      location: [
        {
          key: 'Location',
          value: value,
        },
      ],
    },
  };
}

export const handler: awsLambda.CloudFrontRequestHandler = (
  event: awsLambda.CloudFrontRequestEvent,
  _context: awsLambda.Context,
  callback: awsLambda.Callback<awsLambda.CloudFrontRequestResult>,
): void => {
  const request = event.Records[0].cf.request;

  let prefixPath;

  if (request.uri.match('.+/$')) {
    request.uri += 'index.html';
    callback(null, request);
  } else if ((prefixPath = request.uri.match('(.+)/index.html'))) {
    callback(null, generateRedirectResponse(prefixPath[1] + '/'));
  } else if (request.uri.match('/[^/.]+$')) {
    callback(null, generateRedirectResponse(request.uri + '/'));
  } else {
    callback(null, request);
  }
};
