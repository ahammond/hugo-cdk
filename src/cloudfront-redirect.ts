import { aws_iam, aws_lambda, aws_lambda_nodejs, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class CloudfrontRedirect extends Construct {
  public readonly fn: aws_lambda.Function;
  public readonly version: aws_lambda.Version;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    const role = new aws_iam.Role(this, 'RedirectLambdaRole', {
      assumedBy: new aws_iam.ServicePrincipal('lambda.amazonaws.com'),
    });
    role.assumeRolePolicy?.addStatements(
      new aws_iam.PolicyStatement({
        effect: aws_iam.Effect.ALLOW,
        actions: ['sts:AssumeRole'],
        principals: [new aws_iam.ServicePrincipal('edgelambda.amazonaws.com')],
      }),
    );

    this.fn = new aws_lambda_nodejs.NodejsFunction(this, 'cloudfront-redirect', {
      role: role,
      bundling: {
        minify: true,
        externalModules: [
          'aws-sdk', // Use the 'aws-sdk' available in the Lambda runtime
          '@types/aws-lambda',
        ],
      },
    });

    this.version = new aws_lambda.Version(this, 'RedirectLambdaVersion', {
      lambda: this.fn,
    });
  }
}

export class CloudfrontRedirectStack extends Stack {
  public readonly fn: aws_lambda.Function;
  public readonly version: aws_lambda.Version;

  constructor(scope: Construct, id: string, props: StackProps) {
    super(scope, id, props);

    const c = new CloudfrontRedirect(this, 'RedirectLambdaApplication');
    this.fn = c.fn;
    this.version = c.version;
  }
}
