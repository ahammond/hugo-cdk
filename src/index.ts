export * from './builder';
export * from './cloudfront-redirect';
export * from './hugo-site';
// export * from './standard-redirects-for-cloudfront';
export * from './static-site';
export * from './breaking';
