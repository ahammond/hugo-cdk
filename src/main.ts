#!/usr/bin/env node
import 'source-map-support/register';
import { App } from 'aws-cdk-lib';
import { CloudfrontRedirectStack, HugoSiteStack } from '.';

const commonProps = {
  env: {
    account: '263869919117', // ahammond's AWS account ID. Use your own.
    region: 'us-east-1', // Not sure what other regions will support all this.
  },
};

const app = new App();

// Do not change this at all, even if you have multiple sites in the same account.
// We intentionally want to reuse this lambda for all of them...
// Unless you have billing concerns, but that's better solved by using separate AWS accounts.
const redirect = new CloudfrontRedirectStack(app, 'RedirectLambda', {
  ...commonProps,
  stackName: 'RedirectLambda', // Be super explicit with CDK about the name of this stack.
});

const siteProps = {
  ...commonProps,
  githubOrg: 'ahammond',
  redirectFn: redirect.version,
  siteDomain: 'agh1973.com',
};

// You can just add stanza after stanza to implement more sites.
new HugoSiteStack(app, 'Blog', {
  ...siteProps,
  siteName: 'blog',
});

new HugoSiteStack(app, 'Food', {
  ...siteProps,
  siteName: 'food',
});

new HugoSiteStack(app, 'Ferries', {
  ...siteProps,
  siteName: 'ferries',
});
