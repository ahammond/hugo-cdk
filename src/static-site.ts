import {
  aws_certificatemanager,
  aws_cloudfront,
  aws_cloudfront_origins,
  aws_iam,
  aws_lambda,
  aws_s3,
  aws_route53,
  aws_route53_targets,
  Duration,
  RemovalPolicy,
  Stack,
  StackProps,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface StaticSiteProps {
  /**
   * The name of your site: "myblog".mydomain.com
   *
   * Obviously this needs to be lower case and no special characters.
   * See RFC-1034 and RFC-1035 for details.
   */
  readonly siteName: string;
  /**
   * The domain of your blog: mydomain.com
   * This domain name MUST be part of a zone managed by AWS Route53.
   * It's easiest to just use AWS as your registrar,
   * but either way the zone has to be in AWS Route53 for it to work.
   */
  readonly siteDomain: string;
  /**
   * The standard-redirects-for-cloudfront lambda
   */
  readonly redirectFn: aws_lambda.IVersion;
}

export interface IStaticSite {
  readonly bucket: aws_s3.IBucket;
  readonly distribution: aws_cloudfront.IDistribution;
}

/**
 * Generate a static website distribution in AWS.
 * Includes:
 * - S3 bucket to host content
 * - ACM certificate for https (dns validated)
 * - CloudFront CDN for secure delivery of content
 * - OAI for secure access to the S3 bucket
 * - basic XSS protections
 * - DNS management for the CloudFront domain
 */
export class StaticSite extends Construct implements IStaticSite {
  public readonly bucket: aws_s3.Bucket;
  public readonly distribution: aws_cloudfront.Distribution;

  constructor(scope: Construct, id: string, props: StaticSiteProps) {
    super(scope, id);

    const fqdn = `${props.siteName}.${props.siteDomain}`;
    const zone = aws_route53.HostedZone.fromLookup(this, 'Zone', { domainName: props.siteDomain });

    const certificate = new aws_certificatemanager.DnsValidatedCertificate(this, 'Certificate', {
      domainName: fqdn,
      hostedZone: zone,
    });

    const logBucket = new aws_s3.Bucket(this, 'LogBucket', {
      bucketName: `${fqdn}-logs`,
      encryption: aws_s3.BucketEncryption.S3_MANAGED,
      removalPolicy: RemovalPolicy.DESTROY,
      accessControl: aws_s3.BucketAccessControl.PRIVATE,
      lifecycleRules: [
        {
          expiration: Duration.days(10), // Do you really want to keep logs for longer than 10 days?
        },
      ],
    });

    // Provide an identity for CloudFront to use to access the S3 bucket.
    const originAccessIdentity = new aws_cloudfront.OriginAccessIdentity(this, 'OriginAccessIdentity', {
      comment: `OAI for ${fqdn}`,
    });

    this.bucket = new aws_s3.Bucket(this, 'Bucket', {
      bucketName: fqdn,
      encryption: aws_s3.BucketEncryption.S3_MANAGED, // https://aws.amazon.com/premiumsupport/knowledge-center/s3-website-cloudfront-error-403/
      removalPolicy: RemovalPolicy.DESTROY,
      accessControl: aws_s3.BucketAccessControl.PRIVATE,
      blockPublicAccess: aws_s3.BlockPublicAccess.BLOCK_ALL,
      websiteIndexDocument: 'index.html',
      websiteErrorDocument: '404.html',
    });

    // Grant the CloudFront OAI access to the bucket.
    this.bucket.addToResourcePolicy(
      new aws_iam.PolicyStatement({
        actions: ['s3:GetBucket*', 's3:GetObject*', 's3:List*'],
        resources: [this.bucket.bucketArn, `${this.bucket.bucketArn}/*`],
        principals: [
          new aws_iam.CanonicalUserPrincipal(originAccessIdentity.cloudFrontOriginAccessIdentityS3CanonicalUserId),
        ],
      }),
    );

    this.distribution = new aws_cloudfront.Distribution(this, 'Distribution', {
      certificate,
      defaultBehavior: {
        origin: new aws_cloudfront_origins.S3Origin(this.bucket, {
          originAccessIdentity,
        }),
        edgeLambdas: [
          {
            functionVersion: props.redirectFn,
            eventType: aws_cloudfront.LambdaEdgeEventType.ORIGIN_REQUEST,
          },
        ],
        viewerProtocolPolicy: aws_cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
      },
      domainNames: [fqdn],
      errorResponses: [
        {
          httpStatus: 404,
          responseHttpStatus: 200,
          responsePagePath: '/404.html',
        },
      ],
      logBucket,
      priceClass: aws_cloudfront.PriceClass.PRICE_CLASS_100,
    });

    // Route53 alias record for the CloudFront distribution
    new aws_route53.ARecord(this, 'SiteAlias', {
      target: aws_route53.RecordTarget.fromAlias(new aws_route53_targets.CloudFrontTarget(this.distribution)),
      zone: zone,
      recordName: fqdn,
    });
  }
}

export interface StaticSiteStackProps extends StaticSiteProps, StackProps {}

export class StaticSiteStack extends Stack implements IStaticSite {
  public readonly bucket: aws_s3.Bucket;
  public readonly distribution: aws_cloudfront.Distribution;

  constructor(scope: Construct, id: string, props: StaticSiteStackProps) {
    super(scope, id, props);

    const s = new StaticSite(this, id, props);
    this.bucket = s.bucket;
    this.distribution = s.distribution;
  }
}
