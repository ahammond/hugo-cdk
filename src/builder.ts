import { aws_codebuild, aws_iam, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as staticSite from './static-site';

const hugoDeb = 'https://github.com/gohugoio/hugo/releases/download/v0.74.2/hugo_0.74.2_Linux-64bit.deb';

export interface SiteBuilderProps extends StackProps {
  /**
   * The name of the repo (should match the name of the blog, but you do you.)
   */
  githubRepo: string;
  /**
   * The name of the github owner / organization
   */
  githubOrg: string;
  /**
   * Where should master builds be published?
   */
  production: staticSite.IStaticSite;
  /**
   * Where should PR builds be published?
   * @default - Do not publish PR builds, just build them.
   * @todo add support for publishing PR builds.
   */
  //staging?: staticSite.StaticSiteStack;
}

export class SiteBuilder extends Construct {
  public readonly project: aws_codebuild.Project;

  constructor(scope: Construct, id: string, props: SiteBuilderProps) {
    super(scope, id);

    const buildSpec = {
      version: '0.2',
      environment: {
        vars: {
          HUGO_ENV: 'production',
        },
      },
      phases: {
        install: {
          'runtime-versions': {
            python: 3.8,
          },
          'commands': [ // Install hugo
            `curl -L -o hugo.deb "${hugoDeb}"`,
            'dpkg -i hugo.deb',
          ],
        },
        build: {
          commands: [
            // Render the site
            'hugo -v',
          ],
        },
        post_build: {
          commands: [
            // https://docs.aws.amazon.com/codebuild/latest/userguide/build-env-ref-env-vars.html
            'echo "CODEBUILD_WEBHOOK_TRIGGER: $CODEBUILD_WEBHOOK_TRIGGER"',
            `if [ 'branch/master' = "$CODEBUILD_WEBHOOK_TRIGGER" ]; then
              cd public;
              aws s3 sync --delete --no-progress --sse AES256 --storage-class INTELLIGENT_TIERING . "s3://${props.production.bucket.bucketName}";
              aws cloudfront create-invalidation --distribution-id "${props.production.distribution.distributionId}" --paths '/*';
            else
              echo "not master: do not publish";
            fi`,
          ],
        },
      },
    };

    this.project = new aws_codebuild.Project(this, 'Project', {
      projectName: props.githubRepo,
      buildSpec: aws_codebuild.BuildSpec.fromObject(buildSpec),
      source: aws_codebuild.Source.gitHub({
        // Credentials from CLI, search ImportSourceCredentials at https://docs.aws.amazon.com/cdk/api/latest/docs/aws-codebuild-readme.html
        owner: props.githubOrg,
        repo: props.githubRepo,
        webhookFilters: [
          // Build all the PRs to master
          aws_codebuild.FilterGroup.inEventOf(
            aws_codebuild.EventAction.PULL_REQUEST_CREATED,
            aws_codebuild.EventAction.PULL_REQUEST_UPDATED,
            aws_codebuild.EventAction.PULL_REQUEST_REOPENED,
          ).andBaseBranchIs('master'),
          // And any change to master.
          aws_codebuild.FilterGroup.inEventOf(aws_codebuild.EventAction.PUSH).andBranchIs('master'),
        ],
      }),
    });

    // Submodules, please.
    const cfnProject = this.project.node.defaultChild as aws_codebuild.CfnProject;
    cfnProject.addPropertyOverride('Source.GitSubmodulesConfig.FetchSubmodules', 'True');

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    props.production.bucket.grantReadWrite(this.project.role!);

    this.project.addToRolePolicy(
      new aws_iam.PolicyStatement({
        resources: [
          `arn:aws:cloudfront::${props.env?.account}:distribution/${props.production.distribution.distributionId}`,
        ],
        actions: ['cloudfront:CreateInvalidation'],
      }),
    );
  }
}

export class SiteBuilderStack extends Stack {
  public readonly project: aws_codebuild.Project;

  constructor(scope: Construct, id: string, props: SiteBuilderProps) {
    super(scope, id, props);
    const p = new SiteBuilder(this, id, props);
    this.project = p.project;
  }
}
