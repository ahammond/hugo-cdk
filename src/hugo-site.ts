import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as builder from './builder';
import * as staticSite from './static-site';

export interface HugoSiteStackProps extends staticSite.StaticSiteProps, StackProps {
  readonly githubOrg: string;
}

export class HugoSiteStack extends Stack {
  constructor(scope: Construct, id: string, props: HugoSiteStackProps) {
    super(scope, id, props);

    const site = new staticSite.StaticSite(this, `${props.siteName}Site`, props);

    new builder.SiteBuilder(this, `${props.siteName}Builder`, {
      ...props,
      production: site,
      githubRepo: props.siteName,
    });
  }
}
