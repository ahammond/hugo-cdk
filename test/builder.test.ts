import { App, aws_cloudfront, aws_s3, Stack } from 'aws-cdk-lib';
import * as CuT from '../src';

test('builder snapshot', () => {
  const app = new App();
  const stack = new Stack(app, 'TestStack');
  // WHEN
  new CuT.SiteBuilder(stack, 'Test', {
    githubOrg: 'testOrg',
    githubRepo: 'testRepo',
    production: {
      bucket: aws_s3.Bucket.fromBucketAttributes(stack, 'Bucket', {
        bucketName: 'testBucket',
      }),
      distribution: aws_cloudfront.Distribution.fromDistributionAttributes(stack, 'Distribution', {
        distributionId: 'testDistributionId',
        domainName: 'testDomainName',
      }),
    },
  });
  // THEN
  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
});
