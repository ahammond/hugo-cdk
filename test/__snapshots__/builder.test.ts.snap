// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builder snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store.",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "TestProject60EBF133": Object {
      "Properties": Object {
        "Artifacts": Object {
          "Type": "NO_ARTIFACTS",
        },
        "EncryptionKey": "alias/aws/s3",
        "Environment": Object {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:1.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "Name": "testRepo",
        "ServiceRole": Object {
          "Fn::GetAtt": Array [
            "TestProjectRoleF76CE658",
            "Arn",
          ],
        },
        "Source": Object {
          "BuildSpec": "{
  \\"version\\": \\"0.2\\",
  \\"environment\\": {
    \\"vars\\": {
      \\"HUGO_ENV\\": \\"production\\"
    }
  },
  \\"phases\\": {
    \\"install\\": {
      \\"runtime-versions\\": {
        \\"python\\": 3.8
      },
      \\"commands\\": [
        \\"curl -L -o hugo.deb \\\\\\"https://github.com/gohugoio/hugo/releases/download/v0.74.2/hugo_0.74.2_Linux-64bit.deb\\\\\\"\\",
        \\"dpkg -i hugo.deb\\"
      ]
    },
    \\"build\\": {
      \\"commands\\": [
        \\"hugo -v\\"
      ]
    },
    \\"post_build\\": {
      \\"commands\\": [
        \\"echo \\\\\\"CODEBUILD_WEBHOOK_TRIGGER: $CODEBUILD_WEBHOOK_TRIGGER\\\\\\"\\",
        \\"if [ 'branch/master' = \\\\\\"$CODEBUILD_WEBHOOK_TRIGGER\\\\\\" ]; then\\\\n              cd public;\\\\n              aws s3 sync --delete --no-progress --sse AES256 --storage-class INTELLIGENT_TIERING . \\\\\\"s3://testBucket\\\\\\";\\\\n              aws cloudfront create-invalidation --distribution-id \\\\\\"testDistributionId\\\\\\" --paths '/*';\\\\n            else\\\\n              echo \\\\\\"not master: do not publish\\\\\\";\\\\n            fi\\"
      ]
    }
  }
}",
          "GitSubmodulesConfig": Object {
            "FetchSubmodules": "True",
          },
          "Location": "https://github.com/testOrg/testRepo.git",
          "ReportBuildStatus": true,
          "Type": "GITHUB",
        },
        "Triggers": Object {
          "FilterGroups": Array [
            Array [
              Object {
                "Pattern": "PULL_REQUEST_CREATED, PULL_REQUEST_UPDATED, PULL_REQUEST_REOPENED",
                "Type": "EVENT",
              },
              Object {
                "Pattern": "refs/heads/master",
                "Type": "BASE_REF",
              },
            ],
            Array [
              Object {
                "Pattern": "PUSH",
                "Type": "EVENT",
              },
              Object {
                "Pattern": "refs/heads/master",
                "Type": "HEAD_REF",
              },
            ],
          ],
          "Webhook": true,
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "TestProjectRoleDefaultPolicy1B7D287F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "TestProject60EBF133",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "TestProject60EBF133",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":report-group/",
                    Object {
                      "Ref": "TestProject60EBF133",
                    },
                    "-*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::testBucket",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::testBucket/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "cloudfront:CreateInvalidation",
              "Effect": "Allow",
              "Resource": "arn:aws:cloudfront::undefined:distribution/testDistributionId",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestProjectRoleDefaultPolicy1B7D287F",
        "Roles": Array [
          Object {
            "Ref": "TestProjectRoleF76CE658",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestProjectRoleF76CE658": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
